### ISSUE TITLE
false positive for public method return types for empty-paren method just calling a constructor

### ISSUE THREAD

**Issue by mccartney** [scalastyle/scalastyle#294]
The following code raises a Scalastyle violation:
```
class classOK {
  def newMap() = new scala.collection.mutable.HashMap()
}
```

`public.methods.have.type` / `Public method must have explicit type`

while I'd argue it shouldn't. The following syntax is often used when defining Spring beans.

**Comment by matthewfarwell** [comment]
Hi,

The idea behind this check is that when you define a public method, you define an API. If you're defining an API, then you should think carefully about the return type of the method.

So, ignoring for a minute your use case, I would ask if this method newMap() should be returning a HashMap, which is a particular implementation of the Map interface. I would argue that it should be returning a Map, rather than a HashMap. If you were to change it to a TreeMap, then you would be changing the interface. So I would say that the check is doing its job.

In your use case, (if I understand correctly), you are using this convention with new Xxx() specifically for Spring beans. You've got a number of options:

1) Don't use this rule
2) Use a private/protected method, IIRC you can use private constructors/methods in Spring, but it's a long time since I've Spring.
3) Switch off this rule for this class, using // scalastyle:off public.methods.have.type
4) Switch off this rule for this line, using // scalastyle:ignore public.methods.have.type

**Comment by mccartney** [comment]
Thanks for coming back with your comments.
I understand your perspective. I just felt it's redundant to require to type:
```def name(): SomeType = new SomeType```
but I get your argument that in general these two types might not be the same.

Thanks for the suggestions. Obviously we'll be using one of them. 

I am closing the issue and the linked PR.

**Comment by mccartney** [comment]
And @matthewfarwell on a related note, what's your take on this offending method (same checker):
```
object SomeLogic {
   def apply(name: String) = {
      println("OK") 
   }
}
```

Do you think the checker should have a special case for `apply` methods?

**Comment by matthewfarwell** [comment]
For me, same thing. It is often the case that apply() in an object returns an instance of a different class, for example:

```
object Seq {
  def apply[T](): Seq[T] = List()
}
```

so the same applies.

**Comment by mccartney** [comment]
OK. Thanks. I appreciate your prompt responses.
